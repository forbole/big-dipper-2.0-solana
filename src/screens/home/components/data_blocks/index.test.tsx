import React from 'react';
import renderer from 'react-test-renderer';
import {
  MockTheme, wait,
} from '@tests/utils';
import { ApolloProvider } from '@apollo/client';
import {
  createMockClient, createMockSubscription,
} from 'mock-apollo-client';
import {
  LatestBlockHeightListenerDocument,
  AverageBlockTimeDocument,
  TokenPriceListenerDocument,
  ActiveValidatorCountDocument,
} from '@graphql/types';
import DataBlocks from '.';

// ==================================
// mocks
// ==================================
const mockI18n = {
  t: (key: string) => key,
  lang: 'en',
};
jest.mock('next-translate/useTranslation', () => () => mockI18n);
jest.mock('./components', () => ({
  SingleBlock: (props) => <div id="SingleBlock" {...props} />,
}));

const mockLatestBlockHeight = {
  data: {
    height: [
      {
        slot: 953992,
      },
    ],
  },
};

const mockAverageBlockTime = jest.fn().mockResolvedValue({
  data: {
    averageSlotTimePerHour: [
      {
        averageTime: 0.5512172714745062,
      },
    ],
  },
});

const mockTokenPrice = {
  data: {
    tokenPrice: [
      {
        price: 223.65,
        timestamp: '2021-12-02T06:23:51.967',
        marketCap: 68371260917,
        unitName: 'sol',
      },
    ],
  },
};

const mockActiveValidatorsCount = jest.fn().mockResolvedValue({
  data: {
    activeTotal: {
      aggregate: {
        count: 200,
      },
    },
    total: {
      aggregate: {
        count: 532,
      },
    },
  },
});

// ==================================
// unit tests
// ==================================
describe('screen: Home/DataBlocks', () => {
  it('matches snapshot', async () => {
    const mockClient = createMockClient();
    const mockSubscription = createMockSubscription();
    const mockSubscriptionTwo = createMockSubscription();

    mockClient.setRequestHandler(
      LatestBlockHeightListenerDocument,
      () => mockSubscription,
    );

    mockClient.setRequestHandler(
      TokenPriceListenerDocument,
      () => mockSubscriptionTwo,
    );

    mockClient.setRequestHandler(
      AverageBlockTimeDocument,
      mockAverageBlockTime,
    );

    mockClient.setRequestHandler(
      ActiveValidatorCountDocument,
      mockActiveValidatorsCount,
    );

    let component;

    renderer.act(() => {
      component = renderer.create(
        <ApolloProvider client={mockClient}>
          <MockTheme>
            <DataBlocks />
          </MockTheme>
        </ApolloProvider>,
      );
    });
    await wait();

    renderer.act(() => {
      mockSubscription.next(mockLatestBlockHeight);
    });

    await wait();

    renderer.act(() => {
      mockSubscriptionTwo.next(mockTokenPrice);
    });

    await wait();

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    expect(mockActiveValidatorsCount).toBeCalledTimes(1);
    expect(component.root.findByProps({ label: 'averageSlotTime' }).props.value).toEqual('0.55 s');
    expect(component.root.findByProps({ label: 'latestSlot' }).props.value).toEqual('953,992');
    expect(component.root.findByProps({ label: 'price' }).props.value).toEqual('$223.65');
    expect(component.root.findByProps({ description: 'outOfValidators' }).props.value).toEqual('200');
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
});
