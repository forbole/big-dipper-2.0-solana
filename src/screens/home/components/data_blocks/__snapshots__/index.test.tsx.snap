// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`screen: Home/DataBlocks matches snapshot 1`] = `
<div
  className="makeStyles-root"
>
  <div
    className="makeStyles-slotHeight"
    id="SingleBlock"
    label="latestSlot"
    value="953,992"
  />
  <div
    className="makeStyles-slotTime"
    id="SingleBlock"
    label="averageSlotTime"
    value="6.20 s"
  />
  <div
    className="makeStyles-price"
    id="SingleBlock"
    label="price"
    value="$223.65"
  />
  <div
    className="makeStyles-validators"
    description="outOfValidators"
    id="SingleBlock"
    label="activeValidators"
    value="200"
  />
</div>
`;
